hasktorch/src/Torch/Typed/Vision.hs

```diff
@@ -140,3 +162,372 @@ initMnist path = do
   testImagesBS <- decompressFile path "t10k-images-idx3-ubyte.gz"
   testLabelsBS <- decompressFile path "t10k-labels-idx1-ubyte.gz"
   return (MnistData imagesBS labelsBS, MnistData testImagesBS testLabelsBS)
+
+
+-- ImageFolder dataset 
+
+newtype ImageFolder (m :: Type -> Type) (device :: (D.DeviceType, Nat) ) (finalShape :: [Nat]) = ImageFolder { folderData :: FolderData }
+
+
+instance (KnownShape shape, All KnownNat shape, [n, 3, h, w] ~ shape, 1 <= n, KnownDevice device, Applicative m) => 
+  Dataset m (ImageFolder m device shape) Int (Tensor device 'D.Float shape, Tensor device 'D.Int64 '[n]) where
+  getItem ImageFolder{..} ix =  
+    let
+      [n, c, h, w] = shapeVal @shape
+      indexes = [ix * n .. (ix+1) * n - 1]
+      (imgs, labels) = case someShape indexes of
+              SomeShape (Proxy :: Proxy idxs) -> 
+                (getFolderImages @idxs folderData, 
```

Comment:
I'm not sure how this type checks. `SomeShape (Proxy :: Proxy idxs)` is an existentially quantified shape without constraints. Yet you are using `getFolderImages @idxs` with the `'[n, 3, w, h] ~ idxs` constraint here 🤔

Reply:
The shape is still derived by GHC from the instance declaration and type signature of `getItem`, I believe. I had all the type applications in place originally which would've made this more clear, but looked to clean it up before committing. `idxs` represent the actual indicies of the image folder in this case, not the output shape of the tensor.

That's one thing that's probably a holdover from bodging code for myself together. I treat `SomeShape` and `shapeVal` as a general `SomeNatList` and `natListVal` since it will function fine as such. It's the same permute, even though the permute dims aren't exactly a shape, but an ordering of dimensions, I still use `KnownShape` and `shapeVal`. Not great practice in an actual repository, I'll fix it for clarity.